File: 11_动态规划/子数组/最大子数组和/1749. 任意子数组和的绝对值的最大值.py
from typing import List
from itertools import accumulate, pairwise

# 1 <= nums.length <= 105
# 请你找出 nums 中 和的绝对值 最大的任意子数组（可能为空），并返回该 最大值 。
# 前缀和最大最小作差(函数图像上最大差值)


class Solution:
    def maxAbsoluteSum(self, nums: List[int]) -> int:
        preSum = [0] + list(accumulate(nums))
        return max(preSum) - min(preSum)


print(Solution().maxAbsoluteSum(nums=[1, -3, 2, 3, -4]))
# 输出：5
# 解释：子数组 [2,3] 和的绝对值最大，为 abs(2+3) = abs(5) = 5

File: 10_分治法/猫树分治/1749. 任意子数组和的绝对值的最大值.go
package main

const INF int = 1e18

// 1749. 任意子数组和的绝对值的最大值
// https://leetcode.cn/problems/maximum-absolute-sum-of-any-subarray/description/
func maxAbsoluteSum(nums []int) int {
	n := len(nums)
	res := -INF

	var f func(start, end int)
	f = func(start, end int) {
		if start >= end {
			return
		}
		if start+1 == end {
			res = max(res, abs(nums[start]))
			return
		}
		mid := (start + end) >> 1
		f(start, mid) // !不包含mid
		f(mid, end)   // !不包含mid

		// !包含mid
		curSum, preMax, preMin := 0, 0, 0
		for i := mid - 1; i >= start; i-- {
			curSum += nums[i]
			preMax = max(preMax, curSum)
			preMin = min(preMin, curSum)
		}
		curSum, sufMax, sufMin := 0, 0, 0
		for i := mid + 1; i < end; i++ {
			curSum += nums[i]
			sufMax = max(sufMax, curSum)
			sufMin = min(sufMin, curSum)
		}
		res = max(res, abs(nums[mid]+preMax+sufMax))
		res = max(res, abs(nums[mid]+preMin+sufMin))
	}

	f(0, n)
	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}
